ea-2022-04-ndvi-automation.py

```diff
@@ -0,0 +1,645 @@
+# * include only the plot code in the cell identified for the final plot code below
+
+# ## Task 1: 
+# 
+# In the cell below, create a single dataframe containing MEAN NDVI, the site name, 
+# and the date of the data for the HARV site 
+# scene `HARV/landsat-crop/LC080130302017031701T1-SC20181023151837`.  The column names for the  final
+# DataFrame should be`mean_ndvi`, and `site`, and the data should be **indexed on the date**. 
+# 
+# Use the functions that we reviewed in class (or create your own versions of them) to implement your code
+# 
+# ### In the Cell below Place  All Functions Needed to Run this Notebook (20 points)
+
+# In[7]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+# Function A
+def open_clean_bands(band_path,
+                     valid_range=None):
+    """Open and mask a single Landsat band using a pixel_qa layer.
+
+    Parameters
+    ----------
+    band_path : string
+        A path to the array to be opened
+    valid_range : tuple (optional)
+        A tuple of min and max range values for the data. Default = None
+
+    Returns
+    ----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked set to 1 for
+        True (Boolean)
+    """
+    # Open, clip, mask single band using rioxarray to valid range
+    band = (rxr.open_rasterio(band_path, masked=True)
+            .rio.clip(crop_bound.geometry, from_disk=True)
+            .squeeze())
+    
+    # Specify the valid range of values
+    if valid_range: 
+        mask = ((band <= 0) | (band > 10000))
+        band = band.where(~mask, np.nan)
+        
+    return band
+
+# Function B
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   vals):
+    """Open and mask a single Landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        A list containing paths to Landsat bands 4 and 5 as .tif files
+    pixel_qa : xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been turned into a mask (0s and 1s)
+    crop_bounds : geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using
+        rasterio mask()
+    vals : list
+        A list of values needed to create the cloud mask
+
+    Returns
+    --------
+    mean_ndvi : xarray.DataArray
+        A cropped, masked xarray object containing NDVI values
+
+    """
+    # Create empty list
+    bands = []
+    for band_path in all_bands_path:
+        clean_bands = open_clean_bands(band_path=band_path,
+                                valid_range=(0, 10000))
+
+        bands.append(clean_bands)
+
+    # Open and clip cloud mask layer
```

Comment:
This function has nice comments - it's easy to see what's happening. Even better would be notes about why (e.g. why are you choosing to mask after calculating ndvi?)

Reply:
Thanks.